# Tall- og Typekonvertering

**Læringsmål:**

* Datatyper
* Konvertering mellom datatyper
* Funksjoner


## Tutorial: Datatyper

I Python, og andre programmeringsspråk, kan data ha forskjellige _typer_. Forskjellige datatyper egner seg for forskjellige bruksområder. For eksempel hvis vi skal lagre alderen til en person, vil det lønne seg å lagre dette i en `int`. Navnet til samme person, derimot, bør være en `string`. 

Det finnes mange forksjellige datatyper, men vi skal ikke gå igjennom alle her. Det kommer i en senere øving. De du skal lære her er:

* **Integer** - et heltall. F.eks `10`. I Python brukes `int` for en integer
* **Float** - et flyttall (tall med desimal). F.eks `10.5`
* **String** - tekst. F.eks `"ITGK"`. I Python brukes `str` for en string
* **Boolean** - sannhetsverdi. Enten `True` eller `False`. I Python brukes `bool` for boolean
* **List** - en liste med verdier. En liste inneholder variabler/verdier av hvilken som helst datatype. F.eks `[1, 2, "Er ITGK kult?", True]`
* **ndarray**/**np.array** - et array. F.eks `np.array([1,2,3,4])`. 

Les mer om de forksjellige datatypene nedenfor:


### Integer

Integers er enten et negativt heltall, 0 eller et positivt heltall. Som kjent fra matematikken er Integers tallene denotert som $\mathbb{Z}$. (les mer om Integers i matematikken [her](https://en.wikipedia.org/wiki/Integer). La oss nå lage noen ints i Python, det er utrolig lett. Kjør kodeblokken nedenfor:

In [1]:
a = -10
b = 0
c = 10

Integers følger et set med regler, akkurat som i matematikken. Vi kan for eksempel addere integers, hvor resultatet også vil være en integer. Det samme gjelder for multiplikasjon. Utfører vi _divisjon_ med to integers derimot, vil resultatet være en `float`. La oss gjøre litt aritmetiske operasjoner på ints. Prøv å kjøre kodeblokken under:

In [2]:
print(a + b) # Samme som å si -10 + 0
print(b - c) # Samme som 0 - 10
print(a * c) # Samme som -10 * 10
print(b * c) # Samme som 0 * c
print(a / b) # Samme som -10 : 0

-10
-10
-100
0


ZeroDivisionError: division by zero

Whoops, ser du koden over ga en feilmelding? Karer du å se hva feilen er? Hvis ikke er ikke det så farlig, vi forteller deg nå; på siste linje prøver vi å dele på `0`. Dette vet vi fra matematikken at er fyfy, og det samme gjelder i Python. Det som er fint med Python ovenfor matetmatikken er at Python sier ifra når du gjør noe som ikke er lov, slik som over. Det aller verste som kan skje er at programmet kræsjer, og vi må fikse opp i bugs. Se om du klarer å fikse opp i feilen over, slik at programmet kjører uten å kræsje.

### Float

Floats oppfører seg på nesten samme måte som Integers. De består av de rasjonale tallene $\mathbb{Q}$. De skiller seg fra Integers ved at de kan ligge mellom heltall. La oss lage noen floats. Kjør kodeblokken nedenfor:

In [3]:
d = 1.2
e = -4.2
f = 0.0

På samme måte som med `int`s kan vi utføre aritmetiske operasjoner på floats. Kjør kodeblokken under og se at du forstår hva som skjer:

In [4]:
print(c + e) # Samme som 1.2 + (-4.2)
print(c - e) # Samme som 1.2 - (-4.2)
print(f * e) # Samme som 0.0 * (4.2)

5.8
14.2
-0.0


### String

String er en datatype som inneholder tekst. For å lage en streng skriver vi tekst omringet av "fnutter". Vi kan bruke både enkeltfnutter `'Jeg er en streng'`, dobbeltfnutter `"Jeg er en annen streng"` eller trippelfnutter `"""Jeg er enda en streng"""`. Alle tre måtene å skrive strenger på er like riktig, men de har forskjellige bruksområder. Enkelt- og dobbeltfnutter er veldig like. En av forskjellene er at om du bruker enkeltfnutter, kan du ha dobbeltfnutter i teksten uten noe problem, og omvendt ved bruk av dobbeltfnutter. For eksempel `'Ordet "stein" kan være både et navn og et objekt man finner i naturen'` eller `"Ordet 'stein' kan være både et navn og et objekt man finner i naturen"`. Trippeltfnutter lager såkalte "multiline"-strenger. Altså kan vi få strenger på flere linjer. Kjør kodeblokken under og se om du forstår hva som skjer:

In [5]:
s1 = 'Jeg er en streng med enkeltfnutter'
s2 = "Jeg er en streng med dobbeltfnutter"
s3 = """Jeg er en
multiline streng"""

print(s1)
print(s2)
print(s3)

Jeg er en streng med enkeltfnutter
Jeg er en streng med dobbeltfnutter
Jeg er en
multiline streng


Vi kan ikke gjøre aritmetiske operasjoner på strenger, på samme måte som `int`s og `float`s. Det betyr derimot ikke at vi ikke kan bruke matematiske operatorer på strenger:

In [6]:
s4 = '10' + '15' + '20'
print(s4)

101520


`+` operatoren "setter sammen" strenger. Som i eksempelet over setter vi sammen, eller konkatinerer, tre strenger; `'10'`, `'15'` og `'20'`, til én stor streng `'101520'`. Du ser forhåpentligvis at tallene `10`, `15` og `20` _ikke_ blir addert til `45` slik de ville blitt om de var `int`s eller `float`s, men strengene blir konkatinert. 

In [7]:
s5 = '10' * 10
print(s5)

10101010101010101010


`*` operatoren ganger strengen antall ganger. I eksempelet over ganger vi strengen `'10'` med `10`, og får den resulterende strengen `'10101010101010101010'` ('10' 10 ganger), ikke `100` som om vi hadde ganget `int`en `10` med `int`en `10`. Operatorene `-` og `/` kan vi ikke bruke på strenger.

Innimellom er det fint å ha andre datatyper inne i strenger. Dette gjøres lett med **f-strings**:

In [8]:
s6 = f'Jeg er en f-string, og jeg kan ha for eksempel ints i meg: {12345}, eller floats: {123.45}'
print(s6)

Jeg er en f-string, og jeg kan ha for eksempel ints i meg: 12345, eller floats: 123.45


Det som er verdt å merke seg med **f-strings** er at så fort andre datatyper blir inkorporert inne i strengen, er de ikke lenger sin egen datatype. De er nå en del av den nye strengen. F-strings er helt vanlige strenger, men de er litt lettere å formatere de. Inne i krøllparentesene {} kan vi ha stort sett det vi vil, også variabler:

In [9]:
tall = 12345
s7 = f'Her er et tall: {tall}'
print(s7)

Her er et tall: 12345


### Boolean

En `bool` er en sannhetsverdi, enten `True` eller `False`, og er en _veldig_ sentral datatype i programmering. Booleans kan brukes for eksempel til å sjekke om en alder er under eller over `18`.

In [10]:
gjort_oving = False
print(f'Jeg har gjort øvingen min: {gjort_oving}')

Jeg har gjort øvingen min: False


Senere i emnet vil du lære om if-setninger. Da står booleans sentralt. En liten smakebit her:

In [11]:
fint_vaer = True
if fint_vaer:
    print('Det er fint vær, gå ut!')
else:
    print('Det er ikke fint vært, bli inne du')

Det er fint vær, gå ut!


Prøv å endre variabelen `fint_vaer` til `False` og se hva som skjer. Dette er et eksempel på bruk av booleans og hvordan de kan endre utfallet av et program.

### List

Lister er en annen fundamental datatype i Python. Lister er er en samling av verdier, enten av andre datayper, eller av lister selv. For å lage en liste brukes klammeparantesene []. Inne i klammene legger vi verdiene våre, sparert med komma. Prøv å kjøre kodeblokken under, gjerne endre på verdiene også.

In [12]:
liste_med_tall = [1, 2, 3, 4]
print(f'Her har du en liste med tall: {liste_med_tall}')

Her har du en liste med tall: [1, 2, 3, 4]


Du vil lære mer om lister senere, som for eksempel hvordan du henter ut elementer. Det viktigste for nå er å vite hvordan du oppretter en :) Lister kan som sagt inneholde flere forskjellige datatyper:

In [13]:
liste_med_forskjellige_verdier = [1.0, 4, True, 'hei på deg']
print(f'Her har du en liste med forskjellige verdier: {liste_med_forskjellige_verdier}')

Her har du en liste med forskjellige verdier: [1.0, 4, True, 'hei på deg']


På samme måte som strenger kan vi ikke gjøre vanlige matematiske operasjoner på lister. Vi kan derimort gange en liste med et tall, og plusse sammen lister. Oppførselen blir det samme som når vi ganger en streng med et tall, eller plusser sammen to strenger. _Elementente_ i listen blir ikke ganget med tallet, de vil bli replikert X ganger. _Elementene_ i listene vil heller ikke bli plusset sammen ved bruk av `+`, men den ene listen blir lagt til i den andre listen:

In [14]:
liste_med_tall = [1,2,3,4]
liste_med_tall2 = [5,6,7,8]

liste2 = liste_med_tall + liste_med_tall2
print(liste2)

liste3 = liste_med_tall * 10
print(liste3)

[1, 2, 3, 4, 5, 6, 7, 8]
[1, 2, 3, 4, 1, 2, 3, 4, 1, 2, 3, 4, 1, 2, 3, 4, 1, 2, 3, 4, 1, 2, 3, 4, 1, 2, 3, 4, 1, 2, 3, 4, 1, 2, 3, 4, 1, 2, 3, 4]


### Ndarray

Datatypen `ndarray` kommer fra det eksterne biblioteket `numpy` og er en type liste. `ndarray` blir ofte omtalt som "Numpy array" eller bare "array". For å opprette et array må vi først _importere_ Numpy biblioteket. Det betyr i praksis at vi henter kode som er skrevet av noen eksterne, slik at vi kan bruke den. Deretter bruker vi funksjonen `array()` i biblioteket for å opprette arrayet vårt.

I eksempelet nedenfor importeres Numpy, og vi kaller importen for `np`. Deretter oppretter vi arrayet med `np.array([])`. Klammeparantesene inne i de vanlige parantesene her er en helt vanlig liste, som definert over. Denne må være med - vi oppretter arrayet utifra en liste. Eksempelet under oppretter bare et tomt array.

In [15]:
import numpy as np

arr = np.array([])
print(arr)

[]


For å opprette et array med verdier legger vi bare en liste med verdier som argument inn i `array()`-funksjonen:

In [16]:
import numpy as np

arr = np.array([1, 2, 3, 4])
print(arr)

[1 2 3 4]


La du merke til at kommaene i arrayet ikke er med i output? Dette karakteriserer et array.

I motsetning til vanlige lister kan ikke arrays inneholde flere forskjellige datatyper. Datatypen arrayet inneholder blir bestemt av det første elementet i listen:

In [17]:
import numpy as np

arr = np.array(['Jeg er en streng', 1, 2, True, False])
print(arr)

['Jeg er en streng' '1' '2' 'True' 'False']


I output ser du at _alle_ elementene har fnutter rundt seg, som viser at de nå er strenger, selvom vi la de inn som `int`s og `bool`s. 

Arrays skiller seg også fra lister når det kommer til aritmetiske operasjoner. I motsetning til lister hvor for eksempel, elementene i listen blir replikert X når vi ganger listen med et tall, vil den aritmetiske operasjonen bli utført _på_ elementene i et array:

In [18]:
import numpy as np

arr = np.array([1, 2, 3, 4, 5])
print(arr + 2) # Legger til 2 på alle elementene i arrayet
print(arr - 2) # Trekker fra 2 fra alle elementene i arrayet
print(arr * 2) # Ganger alle elementer i arrayet med 2
print(arr / 2) # Deler alle elementer i arrayet med 2

[3 4 5 6 7]
[-1  0  1  2  3]
[ 2  4  6  8 10]
[0.5 1.  1.5 2.  2.5]


Med arrayer kan vi tilogmed legge sammen/gange sammen/trekke fra eller dele forskjellige arrayer. En viktig ting når vi gjør dette er at alle array som inngår i operasjonene har like lengde:

In [19]:
import numpy as np

arr1 = np.array([1, 2, 3, 4, 5, 6])
arr2 = np.array([6, 5, 4, 3, 2, 1])

print(arr1 + arr2)
print(arr1 - arr2)
print(arr1 / arr2)
print(arr1 * arr2)

arr3 = np.array([1, 2, 3])
print(arr1 + arr3)

[7 7 7 7 7 7]
[-5 -3 -1  1  3  5]
[0.16666667 0.4        0.75       1.33333333 2.5        6.        ]
[ 6 10 12 12 10  6]


ValueError: operands could not be broadcast together with shapes (6,) (3,) 

Siste `print()` i kodeblokken over gir en feilmelding. Klarer du å se hvorfor? Og klarer du rette opp i det?

## Tutorial: Konvertering mellom datatyper

Vi kan ha ulike typer data, som tekststrenger (f.eks. `"Python"`), heltall (f.eks. `42`), flyttall (f.eks. `9.80`) og sannhetsverdier (`True`, `False`). Ofte kommer vi i situasjoner hvor vi har data av en viss type, men vi trenger samme data bare med en annen type. Da må vi konvertere dataene. Noen vanlige konverteringsfunksjoner:

**`int()`** - konverterer til heltall.
- `int('423')` gir 423 (dvs. tekststrengen blir konvertert til et tall). Virker kun hvis tekststrengen faktisk inneholder et heltall.
- `int(5.69)` gir 5 (dvs. for flyttall blir desimaldelen fjernet)

**`float()`** - konverterer til flyttall
- `float('5.69')` gir 5.69 (tekststreng konvertert til tall)
- `float('5')` gir 5.0, dvs. float() virker på tekststrenger enten de inneholder flyttall eller heltall (men ikke på strenger som er noe annet enn tall)
- `float(5)` gir 5.0

**`str()`** - konverterer til tekststreng
- `str(42)` gir '42'
- `str(5.69)` gir '5.69'
Koden under feiler fordi vi har glemt å konvertere. Kjør den og se hva som skjer.

In [20]:
alder = '13'
alder_mor = '37'
sum_alder = alder + alder_mor

print(f'Gratulerer, til sammen er dere {sum_alder} år!')

sum_alder = int(alder) + int(alder_mor)

print('Gratulerer, til sammen er dere ' + sum_alder + ' år!')

Gratulerer, til sammen er dere 1337 år!


TypeError: can only concatenate str (not "int") to str

Den første feilen viser seg i linjen "Gratulerer..." Summen skulle ha blitt 50 år. Men vi har de to alderne fortsatt bare lagret som tekststrenger. Da betyr `+` å hekte sammen strengene, ikke å gjøre noen addisjon. Altså får vi `'13' + '37'` som blir `'1337'` heller enn `13 + 37` som blir `50`. Her måtte vi ha konvertert fra tekst til tall før vi gjorde addisjonen.

Den andre feilen oppstår i den siste print-setningen. Vi har på linjen over kalkulert rett alder, ved å konvertere `alder` og `alder_mor` til `int`. Problemet nå ligger i at vi prøver å legge sammen en `string` og en `int`. Som feilmeldingen sier; "can only concatenate str (not "int") to str". En mulig løsning er å konvertere `sum_alder` tilbake til `string` nå, slik av vi kan plusse sammen to strenger, eller bruke f-strings. Mulige løsninger vises under:

In [21]:
alder = '13'
alder_mor = '37'
sum_alder = alder + alder_mor

print(f'Gratulerer, til sammen er dere {sum_alder} år!')

sum_alder = int(alder) + int(alder_mor)

print('Gratulerer, til sammen er dere ' + str(sum_alder) + ' år!')
print(f'Gratulerer, til sammen er dere {sum_alder} år!')

Gratulerer, til sammen er dere 1337 år!
Gratulerer, til sammen er dere 50 år!
Gratulerer, til sammen er dere 50 år!


Altså: bruker `int()` i linje 7, dette gjør at vi får heltall i variablene `alder` og `alder_mor` så vi blir i stand til å regne med dem. Bruker deretter `str()` i linje 9 så denne opplysningen kan settes sammen med annen tekst og brukes i `print()`. Dette eksemplet viser dermed både et tilfelle hvor vi har tekst men trenger tall, og ett hvor vi har et tall men trenger tekst. Hvis det er vi trenger et desimaltall på alder (f.eks. `13.5`) vil imidlertid koden over ikke funke. Da måtte vi ha brukt funksjonen `float()` der vi nå har brukt `int()`.

## Oppgave a: obligatorisk

I koden under er det noe feil. Finn feilene og rett opp i de

Rett utskrift skal være:

```python
25
```

In [23]:
def legg_sammen_to_tall(a, b):
    return int(a) + int(b)

legg_sammen_to_tall(10, 15)

25

## Oppgave b: obligatorisk

Lag en funksjon `legg_til_landskode(telefonnummer, landskode)` som tar inn `telefonnummer` (`int`) og `landskode` (`int`) som parametere og returnerer telefonnummetet prefixet med "+", landskode og et mellomrom.

***Skriv koden din i kodeblokken udner***

In [29]:
def legg_til_landskode(telefonnummer, landskode):
    return(f"+{landskode} {telefonnummer}")

Hvis du har gjort alt rett, skal kodeblokken under gi ut:

```python
+47 12345678
+46 87654321
```

In [30]:
telefonnummer1 = 12345678
landskode1 = 47

telefonnummer2 = 87654321
landskode2 = 46

print(legg_til_landskode(telefonnummer1, landskode1))
print(legg_til_landskode(telefonnummer2, landskode2))

+47 12345678
+46 87654321


## Oppgave c: obligatorisk

Kodeblokken nedenfor innheholder noen variabler. Konverter alle til `int`. **Merk**: Det lurer seg kanskje noen feil i koden!

In [47]:
a = '1'
b = True
c = False
d = '1.5'
e = '2,45'

a = int(a)
b = int(b)
c = int(c)
d = int(float(d))
e = int(float(e.replace(',', '.')))

Hvis du har gjort alt rett, skal kodeblokken under skrive ut:

```python
a er nå 1
b er nå 1
c er nå 0
d er nå 1
e er nå 2
```

In [48]:
print(f'a er nå {a}')
print(f'b er nå {b}')
print(f'c er nå {c}')
print(f'd er nå {d}')
print(f'e er nå {e}')

a er nå 1
b er nå 1
c er nå 0
d er nå 1
e er nå 2


## Oppgave d: frivillig

Lag en funksjon `mult_list_with_x(l, x)` som tar inn en liste `l` og skalar `x` som parametere og returnerer en _liste_ hvor alle elementene er multiplisert med `x`.

***Skriv koden din i kodeblokken nedenfor***

In [18]:
import numpy as np

def mult_list_with_x(l, x):
    l = np.array(l) * x
    return list(l)

Hvis funksjonen din er skrevet rett, skal kodeblokken nedenfor gi output:

```python
[2.0, 3.0, 4.0, 5.0, 6.0]
```

In [19]:
liste = [1, 1.5, 2, 2.5, 3]
skalar = 2

mult_list_with_x(liste, skalar)

[2.0, 3.0, 4.0, 5.0, 6.0]

### Hint

Her må du bruke **numpy** og `np.array()`. For å gjøre om fra et array til en liste kan du bruke `list()`. Husk også å importere **numpy** med `import numpy as np`.

## Tutorial del 2: avrunding av flyttall

Ofte har man flyttall, men trenger heltall, f.eks. hvis man skal bruke innebygde Python-funksjoner som krever heltall som argument, eller skal bruke tallet som indeks til en streng eller liste (som vi vil se senere i pensum). Flyttall kan konverteres til heltall med funksjoner som `int()` eller `round()`. Kodeblokka under viser litt forskjell på hvordan disse virker.

In [20]:
print("int() bare kutter desimalene, uansett hvor stor eller liten desimaldelen er:")
print("int(2.25) er", int(2.25))
print("int(2.5) er", int(2.5))
print("int(2.99) er", int(2.99))
print("round() runder av til nærmeste heltall, f.eks.")
print("round(2.25) er", round(2.25))
print("round(2.51) er", round(2.51))
print("Hva hvis tallet er midt mellom to heltall?")
print("round(2.5) er", round(2.5))
print("round(3.5) er", round(3.5))
print("round() bruker en IEEE standard som velger partallet for midt-imellom-situasjoner.")
print("Mens int() alltid gir heltall kan round() brukes for antall desimaler:")
print("round(2.5488, 1) blir", round(2.5488, 1))
print("round(2.5488, 3) blir", round(2.5488, 3))
print("Med negativt antall desimaler kan vi få round() til å runde større enn heltall:")
print("round(12345.67, -3) blir", round(12345.67, -3))

int() bare kutter desimalene, uansett hvor stor eller liten desimaldelen er:
int(2.25) er 2
int(2.5) er 2
int(2.99) er 2
round() runder av til nærmeste heltall, f.eks.
round(2.25) er 2
round(2.51) er 3
Hva hvis tallet er midt mellom to heltall?
round(2.5) er 2
round(3.5) er 4
round() bruker en IEEE standard som velger partallet for midt-imellom-situasjoner.
Mens int() alltid gir heltall kan round() brukes for antall desimaler:
round(2.5488, 1) blir 2.5
round(2.5488, 3) blir 2.549
Med negativt antall desimaler kan vi få round() til å runde større enn heltall:
round(12345.67, -3) blir 12000.0


Som du ser i eksemplet, blir 2.5 rundet av til 2 mens 3.5 blir rundet til 4. Dette kan virke litt uvant, i dagliglivet er man mest kjent med såkalt "kjøpmannsavrunding", hvor det alltid rundes opp hvis man er midt mellom (dvs., 2.5 skulle i så fall ha blitt rundet til 3). Konsekvent runding oppover når man er midt mellom har imidlertid en uheldig side, nemlig at man pådrar seg en systematisk feil hvis man har mange data som avrundes. Tenk f.eks. temperaturmålinger for lange perioder, hvor man deretter skal regne ut et snitt for hele perioden. Hvis alle temperaturer som er midt når det gjelder siste brukte siffer, rundes opp, vil snittet for perioden alltid bli litt for høyt. Hvis man i stedet går i partallsretning i alle slike midt mellom situasjoner, vil man runde opp cirka halvparten av gangene og ned cirka halvparten av gangene og dermed unngå slike systematiske feil. Men for kjøpmannen er systematisk runding oppover selvsagt bedre med tanke på å få inn mest mulig penger.

I tillegg til `int()` og `round()` kan f-strenger "innebygd" runde av flyttall:

In [21]:
print(f'1.2345 avrundet til 2 desimaler er: {1.2345:.2f}')
print(f'5.4321 avrundet til 0 desimaler er: {5.4321:.0f}')

1.2345 avrundet til 2 desimaler er: 1.23
5.4321 avrundet til 0 desimaler er: 5


Det som skjer her inne i krøllparentesene her er; `1.2345` er tallet vi ønsker runde av, `:` sier "rund av det som står til venstre til det som står til høyre", `.2` sier "gi meg 2 desimaler" og `f` sier at typen skal være `float`. Det som er verdt å merke seg er at denne måten å runde av tall på gir deg ikke muligheten til å bruke tallet videre. Tallet er da inkorporert i strengen. Med `round()` og `int()` kan vi bruke det avrundede tallet videre.

## Oppgave e: obligatorisk

Lag en funksjon `rund_av(tall, desimaler)` som tar inn et tall `tall` som skal avrundes og `desimaler` antall desimaler tallet skal avrundes til som parametere og returnerer det avrundede tallet.

***Skriv koden din i kodeblokken under.***

In [22]:
def rund_av(tall, desimaler):
    return round(tall, desimaler)

Hvis funksjonen din er skrevet rett, skal kodeblokken under gi følgende output:

```python
1.23
1000.0
```

In [23]:
print(rund_av(1.23456, 2))
print(rund_av(1234.5432, -3))

1.23
1000.0
